src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala

```diff
@@ -201,58 +202,82 @@ object PrecompiledContracts {
+    }
 
-      val adjExpLen = adjExpLength(expBytes, expLength)
+    object PostEIP198Cost {
+      private val GQUADDIVISOR = 20
 
-      multComplexity * math.max(adjExpLen, 1) / GQUADDIVISOR
-    }
+      def calculate(baseLength: Int, modLength: Int, expLength: Int, expBytes: ByteString): BigInt = {
+        val multComplexity = getMultComplexity(math.max(baseLength, modLength))
+        val adjExpLen = adjExpLength(expBytes, expLength)
+        multComplexity * math.max(adjExpLen, 1) / GQUADDIVISOR
+      }
 
-    private def adjExpLength(expBytes: ByteString, expLength: Int): Long = {
-      val expHead =
-        if (expLength <= lengthBytes)
-          expBytes.padToByteString(expLength, 0.toByte)
+      private def getMultComplexity(x: BigInt): BigInt = {
+        val x2 = x * x
+        if (x <= 64)
+          x2
+        else if (x <= 1024)
+          x2 / 4 + 96 * x - 3072
         else
-          expBytes.take(lengthBytes).padToByteString(lengthBytes, 0.toByte)
+          x2 / 16 + 480 * x - 199680
+      }
+    }
 
-      val highestBitIndex = math.max(ByteUtils.toBigInt(expHead).bitLength - 1, 0)
+    object PostEIP2565Cost {
+      private val GQUADDIVISOR = 3
 
-      if (expLength <= lengthBytes) {
-        highestBitIndex
-      } else {
-        8L * (expLength - lengthBytes) + highestBitIndex
+      def calculate(baseLength: Int, modLength: Int, expLength: Int, expBytes: ByteString): BigInt = {
+        val multComplexity = getMultComplexity(math.max(baseLength, modLength))
+        val adjExpLen = adjExpLength(expBytes, expLength)
+        val r = multComplexity * math.max(adjExpLen, 1) / GQUADDIVISOR
+        if (r <= 200) 200
+        else r
       }
-    }
 
-    private def getLength(bytes: ByteString, position: Int): Int = {
-      val start = position * lengthBytes
-      safeInt(ByteUtils.toBigInt(bytes.slice(start, start + lengthBytes)))
+      // ceiling(x/8)^2
+      private def getMultComplexity(x: BigInt): BigInt =
+        ((x + 7) / 8).pow(2)
     }
 
     private def getNumber(bytes: ByteString, offset: Int, length: Int): BigInt = {
       val number = bytes.slice(offset, safeAdd(offset, length)).padToByteString(length, 0.toByte)
       ByteUtils.toBigInt(number)
     }
 
+    private def safeAdd(a: Int, b: Int): Int = {
+      safeInt(BigInt(a) + BigInt(b))
+    }
+
     private def safeInt(value: BigInt): Int =
       if (value.isValidInt)
         value.toInt
       else
         Integer.MAX_VALUE
 
-    private def safeAdd(a: Int, b: Int): Int = {
-      safeInt(BigInt(a) + BigInt(b))
+    private def getLength(bytes: ByteString, position: Int): Int = {
+      val start = position * lengthBytes
+      safeInt(ByteUtils.toBigInt(bytes.slice(start, start + lengthBytes)))
     }
 
-    private def getMultComplexity(x: BigInt): BigInt = {
-      val x2 = x * x
+    private def adjExpLength(expBytes: ByteString, expLength: Int): Long = {
```

Comment:
we could have a more explicit name. I guess that `adj` means adjusted but I am not sure.